### ISSUE TITLE
error on saving RNN layer with recurrent_dropout parameter as saved_model 

### ISSUE THREAD

**Issue by deaputri** [tensorflow/tensorflow#34029]
```
layer_input   = Input(shape=(10, 100)) 
    layer_bi_rnn  = Bidirectional(GRU(units=10,recurrent_dropout=0.2, return_sequences=True))(layer_input)
    layer_dense   = TimeDistributed(Dense(5))(layer_bi_rnn)
    layer_act     = Activation('softmax')(layer_dense)
    model         = Model([layer_input], layer_act)

    model.compile(loss='categorical_crossentropy')
```

save this model as saved_model on `tf-nightly 2.1.0-dev20191104` raise bug as : 

```
Attempted to save a function b'__inference_forward_lstm_1_layer_call_fn_19037' which references a symbolic Tensor Tensor("dropout/mul_1:0", shape=(None, 256), dtype=float32) that is not a simple constant. This is not supported.
```

after trying to change some parameter, i got the conclusion that this issue happens because of `recurrent_dropout` parameter. Any suggestion?

**Comment by oanush** [comment]
@deaputri ,
Looks like the code provided is in complete,Can you share a simple and standalone code to reproduce the issue?Thanks!

**Comment by deaputri** [comment]
```
def create_model():

    layer_input   = Input(shape=(20, 256)) 
    layer_bi_rnn  = Bidirectional(GRU(units=512, dropout=0.2, recurrent_dropout=0.2, return_sequences=True, recurrent_initializer='glorot_uniform'))(layer_input)
    layer_dropout = Dropout(0.2)(layer_bi_rnn)
    layer_dense   = TimeDistributed(Dense(50))(layer_dropout)
    layer_act     = Activation('softmax')(layer_dense)
    model         = Model([layer_input], layer_act)

    model.compile(loss='categorical_crossentropy',
                  optimizer=Adam(0.001))

    return model

model = create_model()
tf.saved_model.save(model, 'testdata/')
```

**Comment by oanush** [comment]
@deaputri ,
Hi, the provided code is not enough to replicate. can you please provide complete code to reproduce ?Thanks!

**Comment by deaputri** [comment]
```
from tensorflow.compat.v2.keras.layers import Bidirectional, GRU, Input, Activation
from tensorflow.compat.v2.keras.layers import TimeDistributed, Dense, Dropout
from tensorflow.compat.v2.keras.optimizers import Adam
from tensorflow.keras import backend as K
from tensorflow.keras.models import Model
import tensorflow as tf

def create_model():

    layer_input   = Input(shape=(20, 256)) 
    layer_bi_rnn  = Bidirectional(GRU(units=512, dropout=0.2, recurrent_dropout=0.2, return_sequences=True, recurrent_initializer='glorot_uniform'))(layer_input)
    layer_dropout = Dropout(0.2)(layer_bi_rnn)
    layer_dense   = TimeDistributed(Dense(50))(layer_dropout)
    layer_act     = Activation('softmax')(layer_dense)
    model         = Model([layer_input], layer_act)

    model.compile(loss='categorical_crossentropy',
                  optimizer=Adam(0.001))

    return model

model = create_model()
tf.saved_model.save(model, 'testdata/')
```

**Comment by oanush** [comment]
@deaputri ,
When tried running the given code for` tf-nightly`, i got the following error. Kindly find the [gist](https://colab.sandbox.google.com/gist/oanush/8eda00767d8043f8415d7911c6eee738/34029.ipynb) of colab.Thanks!

**Comment by BolunHan** [comment]
same error, saving any recurrent model with dropout or recurrent_dropout will raise a error.

**Comment by deaputri** [comment]
@BolunHan i can save the model just using dropout. would u try?

**Comment by deaputri** [comment]
@oanush i'm using `tf-nightly==2.1.0-dev20191104` by using `!pip install tf-nightly==2.1.0-dev20191104`

**Comment by BolunHan** [comment]
@deaputri I just figure out that if you call 
`model.save("some_path_not_ended_with.h5.model") `
adding dropout or recurrent_dropout will cause a error.
however, if your given file path ended with '.h5', everything works just fine.

**Comment by deaputri** [comment]
@BolunHan actually i want to save the model as SavedModel. As the tensorflow documentation on [here](https://www.tensorflow.org/guide/saved_model), the code is `tf.saved_model.save(pretrained_model, "/tmp/mobilenet/1/")`

**Comment by BolunHan** [comment]
@deaputri thx. I will give it a try.
I was using tf.keras to build my recurrent network
`model = tf.keras.Model(inputs=inputs, outputs=outputs)
model.save("given_file_path")`
save model using tf.keras will cause the problem.
I didn't test `tf.saved_model.save()` method

**Comment by deaputri** [comment]
@BolunHan There are lots model type we can use to save. This tensor [doc](https://www.tensorflow.org/tutorials/keras/save_and_load) may help you.

**Comment by gowthamkpr** [comment]
@BolunHan Did @deaputri comment help you solve the problem?

**Comment by gowthamkpr** [comment]
Closing this issue as it has been inactive for more than 3 weeks. Please add additional comments and we can open this issue again. Thanks!

**Comment by tanzhenyu** [comment]
https://github.com/tensorflow/tensorflow/issues/33247

**Comment by NLP-ZY** [comment]
same error! use dropout & recurrent_dropout in GRU, when complete train and use tf.saved_model.save will raise an error. I try to save model directly without train, that will be ok.

**Comment by BolunHan** [comment]
@NLP-ZY Just make sure the save_path of your model ended with ".h5", which solved my own problem.

**Comment by NLP-ZY** [comment]
Thanks! but I should save in tf format to use tf serving.

**Comment by jvishnuvardhan** [comment]
@deaputri This was resolved in `tf-nightly`. Please check the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/172a07741aa76b82fa1a128c08e33f5a/untitled775.ipynb).

I am closing this issue as it was resolved. Please feel free to reopen if the issue persists again. Thanks!
